smartsim/_core/entrypoints/dragon.py

```diff
@@ -119,14 +132,36 @@ def run(
         except zmq.Again:
             logger.error("Could not send response back to launcher.")
 
-        dragon_backend.print_status()
-        dragon_backend.update()
         if not (dragon_backend.should_shutdown or SHUTDOWN_INITIATED):
             logger.debug(f"Listening to {dragon_head_address}")
+            if updater_last_beat <= dragon_backend.last_heartbeat:
+                no_update_steps += 1
+                if no_update_steps >= grace_period:
+                    logger.debug("Restarting updater")
+                    del backend_updater
+                    backend_updater = ContextThread(
+                        name="DragonBackend", daemon=True, target=dragon_backend.update
+                    )
+                    backend_updater.start()
+                    no_update_steps = 0
+            else:
+                updater_last_beat = dragon_backend.last_heartbeat
+                no_update_steps = 0
         else:
-            logger.info("Shutdown has been requested")
+            logger.info("Backend shutdown has been requested")
             break
 
+    try:
+        del backend_updater
+    except Exception:
+        logger.debug("Could not delete backend updater thread")
+
+    if not dragon_backend.frontend_shutdown:
+        logger.info("Frontend will have to be shut down externally")
+        while True:
+            time.sleep(5)
+            logger.info("Waiting for external shutdown")
```

Comment:
I'm not sure i see how this infinite loop exits. Did you intend to check for `dragon_backend.frontend_shutdown` each iteration?

Reply:
Nope, this really waits until the user sends `SIGTERM`. It seems like this is not something *we* need anymore, anyhow.

Reply:
Our shutdown here could be kind of slick and based on zmq. By setting a timeout to like 90s for the heartbeats, we could let the queue trigger the final shutdown.. Is that what you're thinking here?

Reply:
I tried something similar, but then it was a bit confusing to re-use the heartbeat timeout that way, plus the heartbeat would not be sent out at this time.